DataSourceServices.cs

```diff
@@ -223,18 +223,35 @@ public sealed class DataSourceServices:IDataSource {
         /// </summary>
         /// <returns>Instance of TwIdentity.</returns>
         private TwIdentity _GetDSIdentity() {
-            var _asm=this.HandlerType.Assembly;
-            var _ds=_asm.GetCustomAttributes(typeof(DataSourceAttribute), false)[0] as DataSourceAttribute;
-            var _groups=(SupportedGroupsAttribute)Attribute.GetCustomAttribute(this.HandlerType, typeof(SupportedGroupsAttribute),true)??new SupportedGroupsAttribute(TwDG.DS2);
-            var _version=new Version(((AssemblyFileVersionAttribute)Attribute.GetCustomAttribute(_asm, typeof(AssemblyFileVersionAttribute), false)??new AssemblyFileVersionAttribute("1.0.0.0")).Version);
-
+            var asm = HandlerType.Assembly;
+            var ds = asm.GetCustomAttributes(typeof(DataSourceAttribute), false)[0] as DataSourceAttribute;
+            var groups = (SupportedGroupsAttribute) Attribute.GetCustomAttribute(HandlerType, typeof(SupportedGroupsAttribute), true) ?? new SupportedGroupsAttribute(TwDG.DS2);
+            var provider = (DataSourceInfoProviderAttribute) Attribute.GetCustomAttribute(HandlerType, typeof(DataSourceInfoProviderAttribute)) ?? new DataSourceInfoProviderAttribute(typeof(DataSourceInfo));
+            IDataSourceInfo custom = new DataSourceInfo();
+            try
+            {
+                custom = Activator.CreateInstance(provider.Type) as IDataSourceInfo ?? new DataSourceInfo();
```

Comment:
I think that code for creating a provider must be moved to property (to a getter).
If custom provide don't exist that need use default provider (but is not empty provider).